internal/sdkv2provider/data_source_jwks_from_certificate.go

```diff
@@ -0,0 +1,164 @@
+	"math/big"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type Keys struct {
+	Keys []Key `json:"keys"`
+}
+
+type Key struct {
+	Kty    *string  `json:"kty"`
+	X5C    []string `json:"x5c"`
+	N      *string  `json:"n"`
+	E      *string  `json:"e"`
+	Kid    *string  `json:"kid"`
+	X5T256 *string  `json:"x5t#256"`
+	X5Dn   *string  `json:"x5dn"`
+}
+
+func dataSourceJwksFromCertificate() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceJwksFromCertificateRead,
+		Schema:      dataSourceJwksFromCertificateSchema(),
+		Description: `Calculates a JSON Web Key Set from a given certificate.`,
+	}
+}
+
+func dataSourceJwksFromCertificateSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"pem": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: `Requires a pem encoded certificate.`,
+		},
+		"treat_independently": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: `Determines whether to generate JWK for only leaf certificate in a chain and put other certificates into x5c (default) or generate JWKs for each certificate in a chain`,
+		},
+		"jwks": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: `The calculated JWKS`,
+		},
+	}
+}
+
+func dataSourceJwksFromCertificateRead(_ context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	pemString := d.Get("pem").(string)
+	treatIndependently := d.Get("treat_independently").(bool)
+
+	chain := decodePem(pemString)
+
+	var keys []Key
+
+	certificates := parseChain(chain.Certificate)
+
+	if treatIndependently {
+		for _, x509Cert := range certificates {
+			keys = processCertificate(x509Cert, keys, []*x509.Certificate{x509Cert})
+		}
+	} else {
+		leaf := certificates[0]
+		keys = processCertificate(leaf, keys, certificates)
+	}
+
+	jsonResult, err := json.Marshal(Keys{Keys: keys})
+
+	if err != nil {
+		diag.FromErr(err)
+	}
+
+	d.SetId(hex.EncodeToString(jsonResult))
+	return diag.FromErr(d.Set("jwks", string(jsonResult)))
+}
+
+func parseChain(chain [][]byte) []*x509.Certificate {
+	var parsedCertificates []*x509.Certificate
+
+	for _, cert := range chain {
+		x509Cert, err := x509.ParseCertificate(cert)
+		parsedCertificates = append(parsedCertificates, x509Cert)
+		if err != nil {
+			diag.FromErr(err)
+		}
+	}
+	return parsedCertificates
+}
+
+func processCertificate(x509Cert *x509.Certificate, keys []Key, chain []*x509.Certificate) []Key {
+	hash := sha256.New()
+	hash.Write(x509Cert.Raw)
+	kid := base64.StdEncoding.EncodeToString(hash.Sum(nil))
+	kty := x509Cert.PublicKeyAlgorithm.String()
+	pulbicKey := x509Cert.PublicKey.(*rsa.PublicKey)
+
+	e := strings.ReplaceAll(base64.StdEncoding.EncodeToString(big.NewInt(int64(pulbicKey.E)).Bytes()), "=", "")
+	n := strings.ReplaceAll(base64.StdEncoding.EncodeToString(pulbicKey.N.Bytes()), "=", "")
```

Comment:
this is incorrect, please use the [jwk](https://github.com/lestrrat-go/jwx/tree/develop/v2/jwk) library to correctly calculate the key values.

Reply:
Changed